internal/v4/auth.go

```diff
@@ -110,17 +110,26 @@ func (h *Handler) authorizeWithPerms(req *http.Request, read, write []string) er
 	return h.authorize(req, acl)
 }
 
-const usernameAttr = "username"
+const (
+	usernameAttr = "username"
+	groupsAttr   = "groups"
+)
 
 func (h *Handler) checkACLMembership(attrs map[string]string, acl []string) error {
 	user := attrs[usernameAttr]
 	if user == "" {
 		return errgo.New("no username declared")
 	}
+	groups := strings.Fields(attrs[groupsAttr])
 	for _, name := range acl {
```

Comment:
What if we put everyone, user and all the strings in groups as keys of a map, and then just iterate once over acls and check if map["name"] exists? We get out of n^2 algorithm.